core/variable_binary_arithmetic.cpp

```diff
@@ -53,6 +55,29 @@ template <class T1, class T2> Variable plus(const T1 &a, const T2 &b) {
   return transform<arithmetic_and_matrix_type_pairs>(a, b, plus_);
 }
 
+template <class T> class hasReciprocal {
+  static void detect(...);
+  template <class U> static decltype(reciprocal(std::declval<U>())) detect(U);
+
+public:
+  static constexpr bool value =
+      !std::is_same_v<void, decltype(detect(std::declval<T>()))>;
+};
+
+static constexpr auto reciprocal_ = [](const auto a_) {
+  if constexpr (hasReciprocal<std::decay_t<decltype(a_)>>::value)
+    return reciprocal(a_);
+  else
+    return static_cast<decltype(a_)>(1) / a_;
```

Comment:
Why is this fallback required? You are only calling `transform` with `float` and `double`?